### PR TITLE
Add bunch of exclusions to dead code analysis for special methods

### LINKED ISSUE #30377: IDE0051 should not be reported for certain methods
In ASP.NET web app `Application_Start` is marked as unused:

image

### LINKED ISSUE #30886: "Private members are unused" fires on serialization constructors
#### Repro steps

``` VB
    Public Class ProjectReloadedException
        Inherits Exception

        Public Sub New()
            MyBase.New(My.Resources.Designer.PPG_ProjectReloadedSomePropertiesMayNotHaveBeenSet)
        End Sub

        ''' <summary>
        ''' Deserialization constructor.  Required for serialization/remotability support
        '''   (not that we expect this to be needed).
        ''' </summary>
        ''' <param name="Info"></param>
        ''' <param name="Context"></param>
        ''' <remarks>
        '''See .NET Framework Developer's Guide, "Custom Serialization" for more information
        ''' </remarks>
        Private Sub New(Info As SerializationInfo, Context As StreamingContext)
            MyBase.New(Info, Context)
        End Sub

    End Class
```

#### Expected behavior
Severity	Code	Description	Project	Project Rank	File	Line	Category	Suppression State	Tool
Message	IDE0051	Private member 'ProjectReloadedException..ctor' is unused.	Microsoft.VisualStudio.AppDesigner	2	E:\project-system2\src\Microsoft.VisualStudio.AppDesigner\PropPages\ProjectReloadedException.vb	32	Code Quality	Active	Compiler

#### Actual behavior
Should not fire

### LINKED ISSUE #30887: "Private members is unused" fires on ShouldSerializeXXX pattern
#### Repro steps

``` VB
        <VBDescription(My.Resources.Microsoft_VisualStudio_Editors_Designer.ConstantResourceIDs.PPG_WebReferenceNameDescription)>
        <MergableProperty(False)>
        <HelpKeyword("Folder Properties.FileName")>
        Public Property Name() As String
            Get
                Try
                    Return _projectItem.Name
                Catch ex As Exception When Common.ReportWithoutCrash(ex, NameOf(Name), NameOf(WebReferenceComponent))
                    Return String.Empty
                End Try
            End Get
            Set(value As String)
                _projectItem.Name = value
                _page.OnWebReferencePropertyChanged(Me)
            End Set
        End Property

        ' Prevent using Bold Font in the property grid (the same style as other reference)
        Private Function ShouldSerializeName() As Boolean
            Return False
        End Function
```
#### Expected behavior
No warning on ShouldSerializeName()

#### Actual behavior
Severity	Code	Description	Project	File	Line	Suppression State
Message	IDE0051	Private member 'WebReferenceComponent.ShouldSerializeName' is unused.	Microsoft.VisualStudio.Editors	E:\project-system2\src\Microsoft.VisualStudio.Editors\PropPages\WebReferenceComponent.vb	41	Active

ShouldSerializeXXX is a pattern used by controls & things that show up in the property grid that determines whether they are written to InitializeComponent for a WinForms component. The associated FxCop original rule has an exclusion for this.

### PR DESCRIPTION
**Pull request by mavasani**
Fixes #30886: Bail out on special serialization constructors and methods with certain special attributes indicating special usage.

Fixes #30887: Bail out on ShouldSerializeXXX and ResetXXX pattern that was also excluded from legacy FxCop implementation and seems to be common in real world code.

Fixes #30377 : Bail out on methods with signature matching event handlers, which are very common in Web apps.

### CODE REVIEW

File: src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -439,14 +484,76 @@ private bool IsCandidateSymbol(ISymbol memberSymbol)
...
+                                        methodSymbol.IsVirtual ||
+                                        methodSymbol.IsOverride ||
+                                        methodSymbol.IsExtern)
+                                    {
+                                        return false;
+                                    }
+
+                                    // Explicit interface implementations are not referenced explicitly,
+                                    // but are still used.
+                                    if (!methodSymbol.ExplicitInterfaceImplementations.IsEmpty)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs]: thanks for breaking these all up.  it really does help.

File: src/Workspaces/Core/Portable/Shared/Extensions/IMethodSymbolExtensions.cs [hunk 2]
```diff
@@ -357,5 +357,16 @@ public static PredefinedOperator GetPredefinedOperator(this IMethodSymbol symbol
...
+
+        /// <summary>
+        /// Returns true for void returning methods with two parameters, where
+        /// the first parameter is of <see cref="object"/> type and the second
+        /// parameter inherits from or equals <see cref="EventArgs"/> type.
+        /// </summary>
+        public static bool HasEventHandlerSignature(this IMethodSymbol method, INamedTypeSymbol eventArgsType)
+            => eventArgsType != null &&
+               method.Parameters.Length == 2 &&
+               method.Parameters[0].Type.SpecialType == SpecialType.System_Object &&
```
[the remark this answers is not in the document]
Comment (mavasani) [reply src/Workspaces/Core/Portable/Shared/Extensions/IMethodSymbolExtensions.cs]: This one is comparing special types. I have changed the type checks in your previous comments.

**Comment by mavasani** [comment]
Ping @dotnet/roslyn-ide for reviews.

File: src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs [hunk 3]
```diff
@@ -472,6 +579,40 @@ private bool IsEntryPoint(IMethodSymbol methodSymbol)
...
+
+                return methodSymbol.ReturnType.SpecialType == SpecialType.System_Boolean  &&
+                    methodSymbol.Parameters.IsEmpty &&
+                    (IsSpecialMethodWithMatchingProperty("ShouldSerialize") ||
+                     IsSpecialMethodWithMatchingProperty("Reset"));
+
+                // Local functions.
+                bool IsSpecialMethodWithMatchingProperty(string prefix)
+                {
+                    if (methodSymbol.Name.StartsWith(prefix))
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs]: ok.  can we at least have a comment or test taht demonstrates this as intentional behavior?  Thanks!

File: src/Workspaces/Core/Portable/Shared/Extensions/IMethodSymbolExtensions.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Shared/Extensions/IMethodSymbolExtensions.cs]: Oops!  you're correct.  Sorry about that :)

File: src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs [hunk 4]
```diff
@@ -496,6 +603,40 @@ private bool IsEntryPoint(IMethodSymbol methodSymbol)
...
+                    return false;
+                }
+            }
+
+            private bool IsISerializableConstructor(IMethodSymbol methodSymbol)
+                => _iSerializableType != null &&
+                   methodSymbol.Parameters.Length == 2 &&
+                   methodSymbol.Parameters[0].Type.Equals(_serializationInfoType) &&
+                   methodSymbol.Parameters[1].Type.Equals(_streamingContextType) &&
+                   methodSymbol.ContainingType.AllInterfaces.Contains(_iSerializableType);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs]: note: not a huge complaint, but it is interesting that you only check one of your special types for null, but you use three of the special types here.  Seems strange.  I am going to assum that .Equals here is well behaved and will not throw if it gets a null symbol passed in, so i think it's ok.

File: src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mavasani) [reply src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs]: Sure, done.

File: src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs (hunk 4 above)
Comment (mavasani) [reply src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs]: Yes, that is correct.

**Comment by mavasani** [comment]
@jinujoseph for approval

**Pull request merged**